File: samples/msgext-action/nodejs/bots/teamsMessagingExtensionsActionBot.js
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.

require('dotenv').config();
const { TeamsActivityHandler, CardFactory, TeamsInfo, MessageFactory } = require('botbuilder');
const baseUrl = process.env.BaseUrl;

/**
 * TeamsMessagingExtensionsActionBot handles messaging extension actions in Microsoft Teams.
 */
class TeamsMessagingExtensionsActionBot extends TeamsActivityHandler {
    constructor() {
        super();
    }

    /**
     * Handles the submit action from the messaging extension.
     * @param {TurnContext} context - The context object for the turn.
     * @param {MessagingExtensionAction} action - The action object.
     */
    async handleTeamsMessagingExtensionSubmitAction(context, action) {
        switch (action.commandId) {
            case 'createCard':
                return createCardCommand(context, action);
            case 'shareMessage':
                return shareMessageCommand(context, action);
            case 'webView':
                return await webViewResponse(action);
        }
    }

    /**
     * Handles the fetch task action from the messaging extension.
     * @param {TurnContext} context - The context object for the turn.
     * @param {MessagingExtensionAction} action - The action object.
     */
    async handleTeamsMessagingExtensionFetchTask(context, action) {
        switch (action.commandId) {
            case 'webView':
                return empDetails();
            case 'Static HTML':
                return dateTimeInfo();
            default:
                try {
                    const member = await this.getSingleMember(context);
                    return {
                        task: {
                            type: 'continue',
                            value: {
                                card: getAdaptiveCardAttachment(),
                                height: 400,
                                title: `Hello ${member}`,
                                width: 300
                            }
                        }
                    };
                } catch (e) {
                    if (e.code === 'BotNotInConversationRoster') {
                        return {
                            task: {
                                type: 'continue',
                                value: {
                                    card: getJustInTimeCardAttachment(),
                                    height: 400,
                                    title: 'Adaptive Card - App Installation',
                                    width: 300
                                }
                            }
                        };
                    }
                    throw e;
                }
        }
    }

    /**
     * Gets a single member from the conversation.
     * @param {TurnContext} context - The context object for the turn.
     */
    async getSingleMember(context) {
        try {
            const member = await TeamsInfo.getMember(context, context.activity.from.id);
            return member.name;
        } catch (e) {
            if (e.code === 'MemberNotFoundInConversation') {
                await context.sendActivity(MessageFactory.text('Member not found.'));
                return e.code;
            }
            throw e;
        }
    }
}

/**
 * Creates a just-in-time card attachment.
 */
function getJustInTimeCardAttachment() {
    return CardFactory.adaptiveCard({
        actions: [
            {
                type: 'Action.Submit',
                title: 'Continue',
                data: { msteams: { justInTimeInstall: true } }
            }
        ],
        body: [
            {
                text: 'Looks like you have not used Action Messaging Extension app in this team/chat. Please click **Continue** to add this app.',
                type: 'TextBlock',
                wrap: true
            }
        ],
        type: 'AdaptiveCard',
        version: '1.0'
    });
}

/**
 * Creates an adaptive card attachment.
 */
function getAdaptiveCardAttachment() {
    return CardFactory.adaptiveCard({
        actions: [{ type: 'Action.Submit', title: 'Close' }],
        body: [
            {
                text: 'This app is installed in this conversation. You can now use it to do some great stuff!!!',
                type: 'TextBlock',
                isSubtle: false,
                wrap: true
            }
        ],
        type: 'AdaptiveCard',
        version: '1.0'
    });
}

/**
 * Handles the create card command.
 * @param {TurnContext} context - The context object for the turn.
 * @param {MessagingExtensionAction} action - The action object.
 */
function createCardCommand(context, action) {
    const data = action.data;
    const heroCard = CardFactory.heroCard(data.title, data.text);
    heroCard.content.subtitle = data.subTitle;
    const attachment = { contentType: heroCard.contentType, content: heroCard.content, preview: heroCard };

    return {
        composeExtension: {
            type: 'result',
            attachmentLayout: 'list',
            attachments: [attachment]
        }
    };
}

/**
 * Handles the share message command.
 * @param {TurnContext} context - The context object for the turn.
 * @param {MessagingExtensionAction} action - The action object.
 */
function shareMessageCommand(context, action) {
    let userName = 'unknown';
    if (action.messagePayload.from && action.messagePayload.from.user && action.messagePayload.from.user.displayName) {
        userName = action.messagePayload.from.user.displayName;
    }

    let images = [];
    const includeImage = action.data.includeImage;
    if (includeImage === 'true') {
        images = ['https://encrypted-tbn0.gstatic.com/images?q=tbn:ANd9GcQtB3AwMUeNoq4gUBGe6Ocj8kyh3bXa9ZbV7u1fVKQoyKFHdkqU'];
    }
    const heroCard = CardFactory.heroCard(`${userName} originally sent this message:`, action.messagePayload.body.content, images);

    if (action.messagePayload.attachments && action.messagePayload.attachments.length > 0) {
        heroCard.content.subtitle = `(${action.messagePayload.attachments.length} Attachments not included)`;
    }

    const attachment = { contentType: heroCard.contentType, content: heroCard.content, preview: heroCard };

    return {
        composeExtension: {
            type: 'result',
            attachmentLayout: 'list',
            attachments: [attachment]
        }
    };
}

/**
 * Returns employee details task module.
 */
function empDetails() {
    return {
        task: {
            type: 'continue',
            value: {
                width: 350,
                height: 300,
                title: 'Task module WebView',
                url: `${baseUrl}/customForm`
            }
        }
    };
}

/**
 * Returns date and time info task module.
 */
function dateTimeInfo() {
    return {
        task: {
            type: 'continue',
            value: {
                width: 450,
                height: 125,
                title: 'Task module Static HTML',
                url: `${baseUrl}/staticPage`
            }
        }
    };
}

/**
 * Handles the web view response.
 * @param {MessagingExtensionAction} action - The action object.
 */
async function webViewResponse(action) {
    const data = action.data;
    const heroCard = CardFactory.heroCard(`ID: ${data.EmpId}`, `E-Mail: ${data.EmpEmail}`);
    heroCard.content.subtitle = `Name: ${data.EmpName}`;
    const attachment = { contentType: heroCard.contentType, content: heroCard.content, preview: heroCard };
    return {
        composeExtension: {
            type: 'result',
            attachmentLayout: 'list',
            attachments: [attachment]
        }
    };
}

module.exports.TeamsMessagingExtensionsActionBot = TeamsMessagingExtensionsActionBot;

File: samples/msgext-action/csharp/Bots/TeamsMessagingExtensionsActionBot.cs
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.

using System;
using System.Collections.Generic;
using System.IO;
using System.Threading;
using System.Threading.Tasks;
using AdaptiveCards;
using Microsoft.Bot.Builder;
using Microsoft.Bot.Builder.Teams;
using Microsoft.Bot.Schema;
using Microsoft.Bot.Schema.Teams;
using Microsoft.Extensions.Configuration;
using Newtonsoft.Json;
using Newtonsoft.Json.Linq;
using Microsoft.BotBuilderSamples.Helpers;
using Microsoft.BotBuilderSamples.Models;

namespace Microsoft.BotBuilderSamples.Bots
{
    /// <summary>
    /// TeamsMessagingExtensionsActionBot handles messaging extension actions for Teams.
    /// </summary>
    public class TeamsMessagingExtensionsActionBot : TeamsActivityHandler
    {
        private readonly string baseUrl;

        public TeamsMessagingExtensionsActionBot(IConfiguration configuration) : base()
        {
            this.baseUrl = configuration["BaseUrl"];
        }

        /// <summary>
        /// Handles the submission of messaging extension actions.
        /// </summary>
        protected override async Task<MessagingExtensionActionResponse> OnTeamsMessagingExtensionSubmitActionAsync(
            ITurnContext<IInvokeActivity> turnContext,
            MessagingExtensionAction action,
            CancellationToken cancellationToken)
        {
            return action.CommandId switch
            {
                "createCard" => CreateCardCommand(turnContext, action),
                "shareMessage" => ShareMessageCommand(turnContext, action),
                "webView" => WebViewResponse(turnContext, action),
                "createAdaptiveCard" => CreateAdaptiveCardResponse(turnContext, action),
                "razorView" => RazorViewResponse(turnContext, action),
                "HTML" => ShareHtmlCard(turnContext, action),
                _ => await Task.FromResult(new MessagingExtensionActionResponse())
            };
        }

        private MessagingExtensionActionResponse RazorViewResponse(ITurnContext<IInvokeActivity> turnContext, MessagingExtensionAction action)
        {
            var cardData = JsonConvert.DeserializeObject<RazorViewResponse>(action.Data.ToString());
            var card = new HeroCard
            {
                Title = "Requested User: " + turnContext.Activity.From.Name,
                Text = cardData.DisplayData,
            };

            var attachments = new List<MessagingExtensionAttachment>
                {
                    new MessagingExtensionAttachment
                    {
                        Content = card,
                        ContentType = HeroCard.ContentType,
                        Preview = card.ToAttachment(),
                    }
                };

            return new MessagingExtensionActionResponse
            {
                ComposeExtension = new MessagingExtensionResult
                {
                    AttachmentLayout = "list",
                    Type = "result",
                    Attachments = attachments,
                },
            };
        }

        private MessagingExtensionActionResponse CreateCardCommand(ITurnContext<IInvokeActivity> turnContext, MessagingExtensionAction action)
        {
            var createCardData = ((JObject)action.Data).ToObject<CardResponse>();

            var card = new HeroCard
            {
                Title = createCardData.Title,
                Subtitle = createCardData.Subtitle,
                Text = createCardData.Text,
            };

            var attachments = new List<MessagingExtensionAttachment>
                {
                    new MessagingExtensionAttachment
                    {
                        Content = card,
                        ContentType = HeroCard.ContentType,
                        Preview = card.ToAttachment(),
                    }
                };

            return new MessagingExtensionActionResponse
            {
                ComposeExtension = new MessagingExtensionResult
                {
                    AttachmentLayout = "list",
                    Type = "result",
                    Attachments = attachments,
                },
            };
        }

        private MessagingExtensionActionResponse ShareMessageCommand(ITurnContext<IInvokeActivity> turnContext, MessagingExtensionAction action)
        {
            var heroCard = new HeroCard
            {
                Title = $"{action.MessagePayload.From?.User?.DisplayName} originally sent this message:",
                Text = action.MessagePayload.Body.Content,
            };

            if (action.MessagePayload.Attachments != null && action.MessagePayload.Attachments.Count > 0)
            {
                heroCard.Subtitle = $"({action.MessagePayload.Attachments.Count} Attachments not included)";
            }

            var includeImage = ((JObject)action.Data)["includeImage"]?.ToString();
            if (string.Equals(includeImage, bool.TrueString, StringComparison.OrdinalIgnoreCase))
            {
                heroCard.Images = new List<CardImage>
                    {
                        new CardImage { Url = "https://encrypted-tbn0.gstatic.com/images?q=tbn:ANd9GcQtB3AwMUeNoq4gUBGe6Ocj8kyh3bXa9ZbV7u1fVKQoyKFHdkqU" },
                    };
            }

            return new MessagingExtensionActionResponse
            {
                ComposeExtension = new MessagingExtensionResult
                {
                    Type = "result",
                    AttachmentLayout = "list",
                    Attachments = new List<MessagingExtensionAttachment>
                        {
                            new MessagingExtensionAttachment
                            {
                                Content = heroCard,
                                ContentType = HeroCard.ContentType,
                                Preview = heroCard.ToAttachment(),
                            },
                        },
                },
            };
        }

        private MessagingExtensionActionResponse WebViewResponse(ITurnContext<IInvokeActivity> turnContext, MessagingExtensionAction action)
        {
            var cardData = JsonConvert.DeserializeObject<CustomFormResponse>(action.Data.ToString());
            var imgUrl = baseUrl + "/profile-image.png";

            var card = new ThumbnailCard
            {
                Title = "ID: " + cardData.EmpId,
                Subtitle = "Name: " + cardData.EmpName,
                Text = "E-Mail: " + cardData.EmpEmail,
                Images = new List<CardImage> { new CardImage { Url = imgUrl } },
            };

            var attachments = new List<MessagingExtensionAttachment>
                {
                    new MessagingExtensionAttachment
                    {
                        Content = card,
                        ContentType = ThumbnailCard.ContentType,
                        Preview = card.ToAttachment(),
                    }
                };

            return new MessagingExtensionActionResponse
            {
                ComposeExtension = new MessagingExtensionResult
                {
                    AttachmentLayout = "list",
                    Type = "result",
                    Attachments = attachments,
                },
            };
        }

        private MessagingExtensionActionResponse CreateAdaptiveCardResponse(ITurnContext<IInvokeActivity> turnContext, MessagingExtensionAction action)
        {
            var createCardResponse = ((JObject)action.Data).ToObject<CardResponse>();
            var attachments = CardHelper.CreateAdaptiveCardAttachment(action, createCardResponse);

            return new MessagingExtensionActionResponse
            {
                ComposeExtension = new MessagingExtensionResult
                {
                    AttachmentLayout = "list",
                    Type = "result",
                    Attachments = attachments,
                },
            };
        }

        private MessagingExtensionActionResponse ShareHtmlCard(ITurnContext<IInvokeActivity> turnContext, MessagingExtensionAction action)
        {
            var createCardResponse = ((JObject)action.Data).ToObject<CardResponse>();
            var attachments = CardHelper.CreateAdaptiveCardAttachmentForHtml(action, createCardResponse);

            return new MessagingExtensionActionResponse
            {
                ComposeExtension = new MessagingExtensionResult
                {
                    AttachmentLayout = "list",
                    Type = "result",
                    Attachments = attachments,
                },
            };
        }

        /// <summary>
        /// Handles the fetching of tasks for messaging extensions.
        /// </summary>
        protected override async Task<MessagingExtensionActionResponse> OnTeamsMessagingExtensionFetchTaskAsync(
            ITurnContext<IInvokeActivity> turnContext,
            MessagingExtensionAction action,
            CancellationToken cancellationToken)
        {
            return action.CommandId switch
            {
                "webView" => EmpDetails(turnContext, action),
                "HTML" => TaskModuleHtmlPage(turnContext, action),
                "razorView" => DateDayInfo(turnContext, action),
                _ => await HandleDefaultFetchTaskAsync(turnContext, cancellationToken)
            };
        }

        private async Task<MessagingExtensionActionResponse> HandleDefaultFetchTaskAsync(ITurnContext<IInvokeActivity> turnContext, CancellationToken cancellationToken)
        {
            string memberName;
            try
            {
                var member = await TeamsInfo.GetMemberAsync(turnContext, turnContext.Activity.From.Id, cancellationToken);
                memberName = member.Name;
            }
            catch (ErrorResponseException ex)
            {
                if (ex.Body.Error.Code == "BotNotInConversationRoster")
                {
                    return new MessagingExtensionActionResponse
                    {
                        Task = new TaskModuleContinueResponse
                        {
                            Value = new TaskModuleTaskInfo
                            {
                                Card = GetAdaptiveCardAttachmentFromFile("justintimeinstallation.json"),
                                Height = 200,
                                Width = 400,
                                Title = "Adaptive Card - App Installation",
                            },
                        },
                    };
                }
                throw;
            }

            return new MessagingExtensionActionResponse
            {
                Task = new TaskModuleContinueResponse
                {
                    Value = new TaskModuleTaskInfo
                    {
                        Card = GetAdaptiveCardAttachmentFromFile("adaptiveCard.json"),
                        Height = 200,
                        Width = 400,
                        Title = $"Welcome {memberName}",
                    },
                },
            };
        }

        private MessagingExtensionActionResponse DateDayInfo(ITurnContext<IInvokeActivity> turnContext, MessagingExtensionAction action)
        {
            return new MessagingExtensionActionResponse
            {
                Task = new TaskModuleContinueResponse
                {
                    Value = new TaskModuleTaskInfo
                    {
                        Height = 175,
                        Width = 300,
                        Title = "Task Module Razor View",
                        Url = baseUrl + "/Home/RazorView",
                    },
                },
            };
        }

        private MessagingExtensionActionResponse TaskModuleHtmlPage(ITurnContext<IInvokeActivity> turnContext, MessagingExtensionAction action)
        {
            return new MessagingExtensionActionResponse
            {
                Task = new TaskModuleContinueResponse
                {
                    Value = new TaskModuleTaskInfo
                    {
                        Height = 200,
                        Width = 400,
                        Title = "Task Module HTML Page",
                        Url = baseUrl + "/Home/HtmlPage",
                    },
                },
            };
        }

        private MessagingExtensionActionResponse EmpDetails(ITurnContext<IInvokeActivity> turnContext, MessagingExtensionAction action)
        {
            return new MessagingExtensionActionResponse
            {
                Task = new TaskModuleContinueResponse
                {
                    Value = new TaskModuleTaskInfo
                    {
                        Height = 300,
                        Width = 450,
                        Title = "Task Module WebView",
                        Url = baseUrl + "/Home/CustomForm",
                    },
                },
            };
        }

        private static Attachment GetAdaptiveCardAttachmentFromFile(string fileName)
        {
            var adaptiveCardJson = File.ReadAllText(Path.Combine(".", "Resources", fileName));

            return new Attachment
            {
                ContentType = "application/vnd.microsoft.card.adaptive",
                Content = JsonConvert.DeserializeObject(adaptiveCardJson),
            };
        }
    }
}
